File: android/nativebrik/src/main/java/com/nativebrik/sdk/data/experiment.kt
package com.nativebrik.sdk.data

import com.nativebrik.sdk.Config
import com.nativebrik.sdk.schema.ExperimentConfigs
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement

internal interface ExperimentRepository {
    suspend fun fetchExperimentConfigs(id: String): Result<ExperimentConfigs>
    suspend fun fetchTriggerExperimentConfigs(name: String): Result<ExperimentConfigs>
}

internal class ExperimentRepositoryImpl(private val config: Config, private val cache: CacheStore): ExperimentRepository {
    override suspend fun fetchExperimentConfigs(
        id: String
    ): Result<ExperimentConfigs> {
        return withContext(Dispatchers.IO) {
            val url = config.endpoint.cdn + "/projects/" + config.projectId + "/experiments/id/" + id
            val response: String = getRequestWithCache(url, cache, syncDateTime = true).getOrElse {
                return@withContext Result.failure(it)
            }
            val json = Json.decodeFromString<JsonElement>(response)
            val configs = ExperimentConfigs.decode(json) ?: return@withContext Result.failure(FailedToDecodeException())
            return@withContext Result.success(configs)
        }
    }

    override suspend fun fetchTriggerExperimentConfigs(name: String): Result<ExperimentConfigs> {
        return withContext(Dispatchers.IO) {
            val url = config.endpoint.cdn + "/projects/" + config.projectId + "/experiments/trigger/" + name
            val response: String = getRequestWithCache(url, cache, syncDateTime = true).getOrElse {
                return@withContext Result.failure(it)
            }
            val json = Json.decodeFromString<JsonElement>(response)
            val configs = ExperimentConfigs.decode(json) ?: return@withContext Result.failure(FailedToDecodeException())
            return@withContext Result.success(configs)
        }
    }
}

File: Sources/Nativebrik/data/experiment.swift
//
//  experiment.swift
//  Nubrick
//
//  Created by Ryosuke Suzuki on 2024/03/06.
//

import Foundation

protocol ExperimentRepository2 {
    func fetchExperimentConfigs(id: String) async -> Result<ExperimentConfigs, NativebrikError>
    func fetchTriggerExperimentConfigs(name: String) async -> Result<ExperimentConfigs, NativebrikError>
}

class ExperimentRepositoryImpl: ExperimentRepository2 {
    private let config: Config
    private let cache: CacheStore
    init(config: Config, cache: CacheStore) {
        self.config = config
        self.cache = cache
    }
    
    func fetchExperimentConfigs(id: String) async -> Result<ExperimentConfigs, NativebrikError> {
        guard let url = URL(string: self.config.cdnUrl + "/projects/" + self.config.projectId + "/experiments/id/" + id) else {
            return Result.failure(NativebrikError.irregular("Failed to create URL object"))
        }
        
        let data = await getData(url: url, syncDateTime: true, cache: self.cache)
        switch data {
        case .success(let data):
            let decoder = JSONDecoder()
            guard let result = try? decoder.decode(ExperimentConfigs.self, from: data) else {
                return Result.failure(NativebrikError.failedToDecode)
            }
            return Result.success(result)
            
        case .failure(let error):
            return Result.failure(error)
        }
    }
    
    func fetchTriggerExperimentConfigs(name: String) async -> Result<ExperimentConfigs, NativebrikError> {
        guard let url = URL(string: config.cdnUrl + "/projects/" + config.projectId + "/experiments/trigger/" + name) else {
            return Result.failure(NativebrikError.irregular("Failed to create URL object"))
        }
        
        let data = await getData(url: url, syncDateTime: true, cache: self.cache)
        switch data {
        case .success(let data):
            let decoder = JSONDecoder()
            guard let result = try? decoder.decode(ExperimentConfigs.self, from: data) else {
                return Result.failure(NativebrikError.failedToDecode)
            }
            return Result.success(result)
            
        case .failure(let error):
            return Result.failure(error)
        }
    }
}
